File: src/fabricflow/fibc/ofc/ovs.py
# -*- coding: utf-8 -*-

# Copyright (C) 2017 Nippon Telegraph and Telephone Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
# implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""
FIBC mod factory
"""

from fabricflow.fibc.ofc import generic

# pylint: disable=invalid-name

def _mpls_l3_vpn_group(dpath, mod, ofctl):
    """
    MPLS L3 VPN Group
    """
    return generic.mpls_l3_vpn_group(dpath, mod, ofctl, False)

setup_flow = generic.setup_flow
vlan_flow = generic.vlan_flow
termination_mac_flow = generic.termination_mac_flow
mpls1_flow = generic.mpls1_flow
unicast_routing_flow = generic.unicast_routing_flow
bridging_flow = generic.bridging_flow
policy_acl_flow = generic.policy_acl_flow
setup_group = generic.setup_group
l2_interface_group = generic.l2_interface_group
l3_unicast_group = generic.l3_unicast_group
l3_ecmp_group = generic.l3_ecmp_group
mpls_interface_group = generic.mpls_interface_group
mpls_l3_vpn_group = _mpls_l3_vpn_group
mpls_tun1_group = generic.mpls_tun1_group
mpls_swap_group = generic.mpls_swap_group
mpls_ecmp_group = generic.mpls_ecmp_group
l2_unfiltered_interface_group = generic.l2_unfiltered_interface_group
pkt_out = generic.pkt_out
get_port_stats = generic.get_port_stats

File: src/fabricflow/ffctl/ovs.go
// -*- coding: utf-8 -*-

// Copyright (C) 2019 Nippon Telegraph and Telephone Corporation.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
// implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"github.com/spf13/cobra"
)

const (
	ovsBridgeDefault = "dp0"
	ofcAddrDefault   = "tcp:127.0.0.1:6633"
)

func doInitOVS(bridge string, ofc string) error {
	if err := execAndOutput("sudo", "ovs-vsctl", "add-br", bridge); err != nil {
		return err
	}

	if err := execAndOutput("sudo", "ovs-vsctl", "set-controller", bridge, ofc); err != nil {
		return err
	}

	return nil
}

func doCleanOVS(bridge string) error {
	return execAndOutput("sudo", "ovs-vsctl", "del-br", bridge)
}

func doStatusOVS(bridge string) error {
	return execAndOutput("sudo", "ovs-vsctl", "show")
}

func ovsCmd() *cobra.Command {
	var bridge string
	var ofcAddr string

	rootCmd := &cobra.Command{
		Use:   "ovs",
		Short: "OVS commands",
		RunE: func(cmd *cobra.Command, args []string) error {
			return doStatusOVS(bridge)
		},
	}

	rootCmd.PersistentFlags().StringVarP(&bridge, "bridge", "", ovsBridgeDefault, "ovs-bridge name.")
	rootCmd.PersistentFlags().StringVarP(&ofcAddr, "controller", "", ofcAddrDefault, "ovs-controller address.")

	initCmd := &cobra.Command{
		Use:   "init",
		Short: "Initialize OVS",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			return doInitOVS(bridge, ofcAddr)
		},
	}

	rootCmd.AddCommand(initCmd)

	cleanCmd := &cobra.Command{
		Use:   "clean",
		Short: "Clean OVS",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			return doCleanOVS(bridge)
		},
	}

	rootCmd.AddCommand(cleanCmd)

	statusCmd := &cobra.Command{
		Use:   "status",
		Short: "Show status",
		Args:  cobra.NoArgs,
		RunE: func(cmd *cobra.Command, args []string) error {
			return doStatusOVS(bridge)
		},
	}

	rootCmd.AddCommand(statusCmd)

	return rootCmd
}
